text/0000-async_await.md

```diff
@@ -0,0 +1,606 @@
+work.  For this reason, the "matrix" approach in which we have independent
+syntax for generator functions, async functions, and async generator functions,
+seems like a more promising approach.
+
+## "Hot async functions"
+
+As proposed by this RFC, all async functions return immediately, without
+evaluating their bodies at all. As discussed above, this is not convenient for
+use cases in which you have an immediate "initialization" step - those use
+cases need to use a terminal async block, for example.
+
+An alternative would be to have async functions immediately evaluate up until
+their first `await`, preserving their state until then. The implementation of
+this would be quite complicated - they would need to have an additional yield
+point within the `await`, prior to polling the future being awaited,
+conditional on whether or not the await is the first await in the body of the
+future.
+
+A fundamental difference between Rust's futures and those from other languages
+is that Rust's futures do not do anything unless polled. The whole system is
+built around this: for example, cancellation is dropping the future for
+precisely this reason. In contrast, in other languages, calling an async fn
+spins up a future that starts executing immediately. This difference carries
+over to `async fn` and `async` blocks as well, where it's vital that the
+resulting future be *actively polled* to make progress. Allowing for partial,
+eager execution is likely to lead to significant confusion and bugs.
+
+This is also complicated from a user perspective - when a portion of the body
+is evaluated depends on whether or not it appears before all `await`
+statements (which could possibly be macro generated). The use of a terminal
+async block provide a clearer mechanism for distinguishing between the
+immediately evaluated and asynchronously evaluated portions of a future with an
+initialization step.
+
+## Using async/await instead of alternative asynchronicity systems
+
+A final - and extreme - alternative would be to abandon futures and async/await
+as the mechanism for async/await in Rust and to adopt a different paradigm.
+Among those suggested are a generalized effects system, monads & do notation,
+green-threading, and stack-full coroutines.
+
+While it is hypothetically plausible that some generalization beyond
+async/await could be supported by Rust, there has not enough research in this
+area to support it in the near-term. Given our goals for 2018 - which emphasize
+shipping - async/await syntax (a concept available widely in many languages
+which interacts well with our existing async IO libraries) is the most logical
+thing to implement at this stage in Rust's evolution.
+
+## Async blocks vs async closures
+
+As noted in the main text, `async` blocks and `async` closures are closely
+related, and are roughly inter-expressible:
+
+```rust
+// equivalent
+async { ... }
+(async || { ... })()
+
+// equivalent
+async |..| { ... }
+|..| async { ... }
```

Comment:
It should be mentioned that the latter does not support early return. Closures support this. Other kinds of blocks don't support it as well.